### PR TITLE
#24895 Refactoring

### LINKED ISSUE #24895: Fix: Bayesian calculation in Goal to minimize a Metric
### Parent Issue

https://github.com/dotCMS/core/issues/22222

### Problem Statement

We have to type of Goal:

- Goal to minimize for Example Bounce Rate, when you have a Bounce Rate goal you want to minimize the amount of Bounce Rate for a specific page, so in this case each Bounce Rate is a fail.
- Goal to maximize for example Reach Page, when you have a Reach Page Goal you want to Maximixe the amount of times the a specific page is reach, so in this case is Reach Page is a Success.

Right now we are taking the amount from the Goal as a Success no matter the kind of Goal the it is, so in the case of we are taking the fail as success.

### Steps to Reproduce

- Create a Experiment, Add a Bounce Rate, a Variant, and 100% of traffic location.
- Start it.
- Go to the Front end to be assign inside the Experiment.
- Navigate into the Experiment's Page, and Close the Tab.
- Check the Experiment Result, You must have a Bounce Rate inside the Variant that you was assing, and this Variant is going to be the winner.

### Acceptance Criteria

The Winner must be the Variant with less Bounce Rate.

### dotCMS Version

latest in master

### Proposed Objective

Please Select

### Proposed Priority

Please Select

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

_No response_

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

_No response_

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by freddyucv**
### Proposed Changes
* Create Goal concept 

https://github.com/dotCMS/core/pull/24966/files#diff-923c006958022d633318f298905a1204b36334ceeb499f7403cdfcfdd966884bR17

* then Update Goals to use this new Goal class instead of Metric 

https://github.com/dotCMS/core/pull/24966/files#diff-a4fe9c24f0b7e5448acac1336a35233412d3ca8d142b6131e5f3e4d7b008163cR21

* Create diferent type of Goal  

https://github.com/dotCMS/core/pull/24966/files#diff-a4fe9c24f0b7e5448acac1336a35233412d3ca8d142b6131e5f3e4d7b008163cR21

* ALso create different kind of Goal according to the Type of Mteric that they use

https://github.com/dotCMS/core/pull/24966/files#diff-0cf29a124b400ea4641cf22ba930616718e00660e4361eca7bef6911969ffca1R1

https://github.com/dotCMS/core/pull/24966/files#diff-6a38862d40569a021caab6c784c5dbeae4f595e21735bd5aa84084e3dad8bb18R1

* uncouple Analizer from Experiment Concept, Metric must not be dependents from Experiment

https://github.com/dotCMS/core/pull/24966/files#diff-b75692fe81272019be341546b06800bd71b258fe7c302b05512aa2f3ba7b1141R23

* Use the type of Goal to now how take the result get from the Analizer

https://github.com/dotCMS/core/pull/24966/files#diff-1fdf3602f53e848245929cb77e96e83012c452faf87ae01d3e699826cfd22f2eR141

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/experiments/model/GoalFactory.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package com.dotcms.experiments.model;
+
+import static com.dotcms.util.CollectionsUtils.map;
+
+import com.dotcms.analytics.metrics.Metric;
+import com.dotcms.analytics.metrics.MetricType;
+import java.util.Map;
+import java.util.function.Function;
+
+public class GoalFactory {
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/experiments/model/GoalFactory.java:10]: Javadoc

File: dotCMS/src/main/java/com/dotcms/experiments/model/Goal.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import com.liferay.util.StringPool;
+import org.immutables.value.Value;
+
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+@JsonSubTypes({
+        @Type(value = ReachPageGoal.class, name = "ReachPage"),
+        @Type(value = BounceRateGoal.class, name = "BounceRate")
+})
+public abstract class Goal {
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/experiments/model/Goal.java:22]: Javadoc

File: dotCMS/src/main/java/com/dotcms/experiments/model/BounceRateGoal.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package com.dotcms.experiments.model;
+
+import com.dotcms.analytics.metrics.Metric;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class BounceRateGoal extends Goal {
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/experiments/model/BounceRateGoal.java:6]: Javadoc

File: dotCMS/src/main/java/com/dotcms/experiments/model/ReachPageGoal.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package com.dotcms.experiments.model;
+
+import com.dotcms.analytics.metrics.Metric;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.commons.math3.optim.nonlinear.scalar.GoalType;
+
+public class ReachPageGoal extends Goal {
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/experiments/model/ReachPageGoal.java:7]: Javadoc

File: dotCMS/src/main/java/com/dotcms/experiments/model/GoalFactory.java (hunk 1 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/model/GoalFactory.java:16]: https://github.com/dotCMS/core/pull/24966/commits/68d273702f357ed63fbea36560e00476205cb41d#diff-9326677392d3b95e397ff9f80557cd8fdad0cc620a363c18ef35754d2f28be0aL10

File: dotCMS/src/main/java/com/dotcms/experiments/model/Goal.java (hunk 2 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/model/Goal.java:27]: done https://github.com/dotCMS/core/pull/24966/commits/68d273702f357ed63fbea36560e00476205cb41d#diff-6a38862d40569a021caab6c784c5dbeae4f595e21735bd5aa84084e3dad8bb18R7

File: dotCMS/src/main/java/com/dotcms/experiments/model/BounceRateGoal.java (hunk 3 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/model/BounceRateGoal.java:12]: done https://github.com/dotCMS/core/pull/24966/commits/68d273702f357ed63fbea36560e00476205cb41d#diff-0cf29a124b400ea4641cf22ba930616718e00660e4361eca7bef6911969ffca1R6

File: dotCMS/src/main/java/com/dotcms/experiments/model/ReachPageGoal.java (hunk 4 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/experiments/model/ReachPageGoal.java:13]: done https://github.com/dotCMS/core/pull/24966/commits/68d273702f357ed63fbea36560e00476205cb41d#diff-6a38862d40569a021caab6c784c5dbeae4f595e21735bd5aa84084e3dad8bb18R7

File: dotCMS/src/main/java/com/dotcms/experiments/model/BounceRateGoal.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+    public BounceRateGoal(final @JsonProperty("metric")  Metric metric) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/experiments/model/BounceRateGoal.java:14]: JsonCreator?

File: dotCMS/src/main/java/com/dotcms/experiments/model/ReachPageGoal.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/experiments/model/ReachPageGoal.java:14]: JsonCreator

**Review by jdotcms — APPROVED** [review]

**Review by dsilvam — APPROVED** [review]

**Pull request merged**